lib/galaxy/config/__init__.py

```diff
@@ -1201,16 +1204,29 @@ def reindex_tool_search(self):
         self.toolbox_search.build_index(tool_cache=self.tool_cache)
         self.tool_cache.reset_status()
 
-    def _set_enabled_container_types(self):
+    def _set_enabled_container_types(self, app_info_kwargs, mulled_resolution_cache):
         container_types_to_destinations = collections.defaultdict(list)
         for destinations in self.job_config.destinations.values():
             for destination in destinations:
+                self._set_destination_container_finder(destination, app_info_kwargs, mulled_resolution_cache)
                 for enabled_container_type in self.container_finder._enabled_container_types(destination.params):
                     container_types_to_destinations[enabled_container_type].append(destination)
         self.toolbox.dependency_manager.set_enabled_container_types(container_types_to_destinations)
         self.toolbox.dependency_manager.resolver_classes.update(self.container_finder.container_registry.resolver_classes)
         self.toolbox.dependency_manager.dependency_resolvers.extend(self.container_finder.container_registry.container_resolvers)
 
+    def _set_destination_container_finder(self, destination, app_info_kwargs, mulled_resolution_cache):
+        if 'container_resolvers' in destination.params:
+            from galaxy.tool_util.deps import containers
+            from galaxy.tool_util.deps.dependencies import AppInfo
```

Comment:
Amazing work. I really appreciate the test cases and while I don't love the destination ID map pattern - it is certainly workable and I think I started this pattern with job metrics that work a lot the same as this right?

I think we should rework this though so that AppInfo reflects the application state and not destination state. Maybe that means renaming AppInfo or maybe that means storing these config parameters on something else and passing them in beside app_info - but app_info is supposed to be proxy for app and not job destination specific things.

Let me take a look and see if I dig into the interactions if I can figure out something that is clean to accomplish this.

Reply:
Yeah - I think there should be just one ContainerFinder that gets the fixed AppInfo and then it should cache ``ContainerRegistery`` s for each DestinationInfo in ``find_container``. I think that would be a better set of abstractions. We're already bundling up the destination info and such and passing it into that layer at that point - it makes sense to keep the logic there. Also - it would prevent us from storing job destination ID encoded stuff on config when it would be better stored on a component that was specific to jobs (the ContainerFinder).

This would allow truly dynamic job container resolution also I think - the destination IDs wouldn't need to be known ahead of time or defined in the config file.

Reply:
I think https://github.com/galaxyproject/galaxy/commit/939a51d7d7f3c8347e13a08d12ce8674f7f7be44 is a better use of the abstractions and allows those dynamic container registries for destinations without IDs. It keeps AppInfo just as appinfo and it parses destination parameters from the argument we're already sending. Also it doesn't keep this dictionary attached to config.

I didn't test it though... obviously 😅.

Reply:
Thanks. I can provide another structure to pass these in pretty easily if you haven't come up with anything.

Reply:
> https://github.com/galaxyproject/galaxy/commit/939a51d7d7f3c8347e13a08d12ce8674f7f7be44

:shipit: